lib/fss.ex

```diff
@@ -13,19 +13,65 @@ defmodule FSS do
 
   defmodule S3 do
     defmodule Config do
-      defstruct [:region, :access_key_id, :secret_access_key]
+      defstruct [
+        :access_key_id,
+        :region,
+        :secret_access_key,
+        {:endpoint, "amazonaws.com"},
+        :token
+      ]
 
       @type t :: %__MODULE__{
-              region: String.t(),
               access_key_id: String.t(),
-              secret_access_key: String.t()
+              endpoint: String.t(),
+              region: String.t(),
+              secret_access_key: String.t(),
+              token: String.t() | nil
             }
     end
 
     defmodule Entry do
-      defstruct [:url, :config]
+      defstruct [:bucket, :key, :port, :config]
+
+      @type t :: %__MODULE__{
+              bucket: String.t(),
+              key: String.t(),
+              port: pos_integer(),
+              config: Config.t()
+            }
+
+      def config_from_system_env() do
+        %FSS.S3.Config{
+          access_key_id: System.get_env("AWS_ACCESS_KEY_ID"),
+          secret_access_key: System.get_env("AWS_SECRET_ACCESS_KEY"),
+          region: System.get_env("AWS_REGION", System.get_env("AWS_DEFAULT_REGION")),
+          token: System.get_env("AWS_SESSION_TOKEN")
+        }
+      end
+
+      def parse(url, opts \\ []) do
+        opts =
+          opts
+          |> Keyword.validate!([:config])
+          |> Keyword.put_new_lazy(:config, fn -> config_from_system_env() end)
```

Comment:
here is one idea:

```suggestion
        config =
          opts
          |> Keyword.validate!([config: nil])
          |> Keyword.fetch!(:config)
          |> case do
            nil -> config_from_system_env()
            config when is_list(config) -> struct!(config_from_system_env(), config)
            %Config{} -> config
          end
          |> validate_config!() # check there are no nils
```

This way we allow no config to be set, %Config{}, or a keyword list to update config. Then we validate no field is missing due to env vars (or default values). WDYT?

Reply:
Implemented in: [174094a](https://github.com/elixir-explorer/explorer/pull/652/commits/174094a3c270fb2873d1b432daaf76d3e7f07231)
Thanks! :purple_heart: